cumulus_library/studies/discovery/code_detection.py

```diff
@@ -0,0 +1,84 @@
+""" Module for generating encounter codeableConcept table"""
+
+from cumulus_library.base_table_builder import BaseTableBuilder
+from cumulus_library.helper import get_progress_bar, query_console_output
+from cumulus_library.template_sql.templates import get_code_system_pairs
+from cumulus_library.template_sql.utils import (
+    is_codeable_concept_array_populated,
+    is_codeable_concept_populated,
+    is_code_populated,
+)
+
+from cumulus_library.studies.discovery.code_definitions import code_list
+
+
+class CodeDetectionBuilder(BaseTableBuilder):
+    display_text = "Selecting unique code systems..."
+
+    def _check_codes_in_fields(self, code_sources: list[dict], schema, cursor) -> dict:
+        """checks if Coding/CodeableConcept fields are present and populated"""
+
+        with get_progress_bar() as progress:
+            task = progress.add_task(
+                "Discovering available coding systems...",
+                # Each column in code_sources requires at most 3 queries to
+                # detect valid data is in the DB
+                total=len(code_sources),
+            )
+            for code_source in code_sources:
+                if code_source["is_array"]:
+                    code_source["has_data"] = is_codeable_concept_array_populated(
+                        schema,
+                        code_source["table_name"],
+                        code_source["column_name"],
+                        cursor,
+                        allow_partial=False,
+                    )
+                elif code_source["is_bare_coding"]:
+                    code_source["has_data"] = is_code_populated(
+                        schema,
+                        code_source["table_name"],
+                        code_source["column_name"],
+                        cursor,
+                        allow_partial=False,
+                    )
+                else:
+                    code_source["has_data"] = is_codeable_concept_populated(
+                        schema,
+                        code_source["table_name"],
+                        code_source["column_name"],
+                        cursor,
+                        allow_partial=False,
+                    )
+                progress.advance(task)
+        return code_sources
+
+    def prepare_queries(self, cursor: object, schema: str):
+        """Constructs queries related to condition codeableConcept
+
+        :param cursor: A database cursor object
+        :param schema: the schema/db name, matching the cursor
+
+        """
+
+        code_sources = []
+        for code_definiton in code_list:
+            if any(
+                x not in code_definiton.keys() for x in ["table_name", "column_name"]
+            ):
```

Comment:
nit: This is a curious check - you are basically validating that you yourself didn't make a mistake when writing `code_list`? Maybe a unit test makes more sense for that?

Reply:
This is due to the seperation and/or having another user less familiar editing the definitions file - basically a consequence of not using a class.

Reply:
Sure I love the check. No complaints about that. It just struck me as an odd place to do it - more of a unit test than a runtime test in my head. But just a nit - location doesn't matter too much and this isn't performance sensitive code.

Reply:
reminder that the contract we have in this repo is that studies themselves are exempt from unit tests, for ease of use by clinical folks - that's the reason for here rather than a unit test

Reply:
Oh sure but like. This is a study we control, in the main library repo, where there are unit tests. Like we could still use one for us right?

I'm not trying to advocate hard for a unit test in this specific case, who cares. But that distinction is troubling to me -- I'd think while we don't require others to have production-quality full QA practices etc, we'd want them as much as possible for our own sanity. None of our code under studies/ has any unit testing?